# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [0]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
        #TODO: finish me
        # YOUR CODE HERE
        #raise NotImplementedError()
        #My code
        return DualNumber(self.value + other.value, self.dvalue + other.dvalue)        
        
    # TODO: add missing methods
    # YOUR CODE HERE
    #raise NotImplementedError()
    #My code
    def __truediv__(self, other):
      den = other.value**2
      num = other.value * self.dvalue - self.value * other.dvalue
      return DualNumber(self.value / other.value, num / den)
    
    def __sub__(self, other):
      return DualNumber(self.value - other.value, self.dvalue - other.dvalue)
    
    def __pow__(self, other):
      return DualNumber(self.value**other.value, other.value * self.value**(other.value - 1) * self.dvalue
                       + math.log(self.value) * self.value**other.value * other.dvalue)

In [0]:
# Tests

DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [0]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [0]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    # YOUR CODE HERE
    #My code
    return DualNumber(math.cos(x.value), (-1) * math.sin(x.value) * x.dvalue)
    
    #raise NotImplementedError()

def tan(x):
    # YOUR CODE HERE
    #raise NotImplementedError()
    #My code
    return DualNumber(math.tan(x.value), (1 / (math.cos(x.value)**2) * x.dvalue))
    

def exp(x):
    # YOUR CODE HERE
    #raise NotImplementedError()
    #My code
    return DualNumber(math.exp(x.value), math.exp(x.value) * x.dvalue)

In [0]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [0]:
# YOUR CODE HERE
#raise NotImplementedError()
#My code
def z_x(x, y):
  return DualNumber(x.value * y.value + math.sin(x.value), y.value * x.dvalue 
                    + math.cos(x.value) * x.dvalue)
print('z is :', z_x(DualNumber(0.5, 1), DualNumber(4.2, 0)).value)
print('z_x is : ', z_x(DualNumber(0.5, 1), DualNumber(4.2, 0)).dvalue)

z is : 2.579425538604203
z_x is :  5.077582561890373


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

**My answer: **

x = 0.5, y = 4.2         
z = x * y + sin(x) = 0.5 x 4.2 + sin(0.5) = 2.5794255   
∂𝑧/∂𝑥 = y + cos(x) = 4.2 + cos(0.5) = 4.2 + 0.8775825 = 5.0775825

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [0]:
# YOUR CODE HERE
#raise NotImplementedError()
#My code
def z_y(x, y):
  return DualNumber(x.value * y.value + math.sin(x.value), x.value * y.dvalue)
dzdy = z_y(DualNumber(0.5, 0), DualNumber(4.2, 1))
dzdy = dzdy.dvalue

print('dz/dy:', dzdy)

dz/dy: 0.5


**My answer:**

Compute by hand:

dz/dy = x = 0.5

In [0]:
#Tests
assert dzdy
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

**Created function: **

z = exp(x)/cos(x)  - tan(x) + x^y

∂𝑧/∂𝑥 = ((exp(x)*cos(x) + exp(x)*sin(x))/cos(x)^2)  - 1/cos(x)^2 +y*x^(y-1)


In [0]:
# YOUR CODE HERE
#raise NotImplementedError()
#My code
x = DualNumber(2, 1)
y = DualNumber(4, 0)
z = (exp(x) / cos(x)) - tan(x) + x**y

dzdx_hand = (math.exp(2) * (math.cos(2) + math.sin(2)) / ((math.cos(2))**2)) - (1 / (math.cos(2))**2) + 4 * 2**(4 - 1)
print('dzdx computed by hand is: ', dzdx_hand)

dzdx = z.dvalue
print('dzdx from the code is: ', dzdx)

dzdx computed by hand is:  47.267034396082366
dzdx from the code is:  47.267034396082366
